File: esoteric_entropy/report.py
"""Report generator for esoteric-entropy randomness test battery."""

from __future__ import annotations

import os
import platform
import time
from datetime import datetime
from pathlib import Path

import numpy as np

from esoteric_entropy.test_suite import TestResult, run_all_tests, calculate_quality_score


def _grade_icon(grade: str) -> str:
    return {"A": "✅", "B": "✅", "C": "⚠️", "D": "⚠️", "F": "❌"}.get(grade, "❓")


def _pass_icon(passed: bool) -> str:
    return "✅" if passed else "❌"


def generate_source_report(name: str, data: np.ndarray, results: list[TestResult]) -> str:
    """Generate markdown section for a single source."""
    score = calculate_quality_score(results)
    passed = sum(1 for r in results if r.passed)
    total = len(results)
    grade = "A" if score >= 80 else "B" if score >= 60 else "C" if score >= 40 else "D" if score >= 20 else "F"

    lines = [
        f"### {name}",
        f"**Score: {score:.1f}/100** | **Grade: {grade}** | **Passed: {passed}/{total}** | **Samples: {len(data):,}**\n",
        "| Test | Result | Grade | P-Value | Statistic | Details |",
        "|------|--------|-------|---------|-----------|---------|",
    ]

    for r in results:
        p_str = f"{r.p_value:.6f}" if r.p_value is not None else "N/A"
        lines.append(
            f"| {r.name} | {_pass_icon(r.passed)} | {_grade_icon(r.grade)} {r.grade} "
            f"| {p_str} | {r.statistic:.4f} | {r.details} |"
        )
    lines.append("")
    return "\n".join(lines)


def generate_full_report(
    source_results: dict[str, tuple[np.ndarray, list[TestResult]]],
    output_path: str | Path | None = None,
) -> str:
    """Generate complete markdown report for all sources."""
    now = datetime.now()

    # Summary table
    summary_rows = []
    for name, (data, results) in sorted(
        source_results.items(),
        key=lambda x: calculate_quality_score(x[1][1]),
        reverse=True,
    ):
        score = calculate_quality_score(results)
        passed = sum(1 for r in results if r.passed)
        total = len(results)
        grade = "A" if score >= 80 else "B" if score >= 60 else "C" if score >= 40 else "D" if score >= 20 else "F"
        summary_rows.append((name, score, grade, passed, total, len(data)))

    lines = [
        f"# 🔬 Esoteric Entropy — Randomness Test Report",
        f"",
        f"**Generated:** {now.strftime('%Y-%m-%d %H:%M:%S')}",
        f"**Machine:** {platform.node()} ({platform.machine()}, {platform.system()} {platform.release()})",
        f"**Python:** {platform.python_version()}",
        f"**Tests in battery:** {len(source_results[list(source_results.keys())[0]][1]) if source_results else 0}",
        f"",
        f"## Summary",
        f"",
        f"| Rank | Source | Score | Grade | Passed | Samples |",
        f"|------|--------|-------|-------|--------|---------|",
    ]

    for i, (name, score, grade, passed, total, samples) in enumerate(summary_rows, 1):
        lines.append(
            f"| {i} | {name} | {score:.1f} | {_grade_icon(grade)} {grade} "
            f"| {passed}/{total} | {samples:,} |"
        )

    lines.append("")
    lines.append("---")
    lines.append("")
    lines.append("## Detailed Results")
    lines.append("")

    for name, (data, results) in sorted(
        source_results.items(),
        key=lambda x: calculate_quality_score(x[1][1]),
        reverse=True,
    ):
        lines.append(generate_source_report(name, data, results))
        lines.append("---\n")

    report = "\n".join(lines)

    if output_path:
        path = Path(output_path)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(report)

    return report

File: crates/esoteric-cli/src/commands/report.rs
use std::time::Instant;

pub fn run(samples: usize, source_name: Option<&str>, output_path: Option<&str>) {
    // Use make_pool which defaults to fast sources
    let filter = source_name.or(None);
    let pool = super::make_pool(filter);
    let infos = pool.source_infos();

    // Get sources from platform detection, filtered same way
    let all_sources = esoteric_core::platform::detect_available_sources();
    let sources: Vec<_> = if let Some(name) = source_name {
        all_sources
            .into_iter()
            .filter(|s| s.name().to_lowercase().contains(&name.to_lowercase()))
            .collect()
    } else {
        let fast_names: Vec<_> = infos.iter().map(|i| i.name.clone()).collect();
        all_sources
            .into_iter()
            .filter(|s| fast_names.iter().any(|n| n == s.name()))
            .collect()
    };

    if sources.is_empty() {
        if let Some(name) = source_name {
            eprintln!("Source '{name}' not found.");
        } else {
            eprintln!("No sources found.");
        }
        std::process::exit(1);
    }

    println!(
        "🔬 Running full test battery on {} source(s), {} samples each...\n",
        sources.len(),
        samples
    );

    let mut all_results = Vec::new();

    for src in &sources {
        let info = src.info();
        print!("  Collecting from {}...", info.name);

        let t0 = Instant::now();
        let data = src.collect(samples);
        print!(" {} bytes", data.len());

        if data.is_empty() {
            println!(" ✗ no data");
            continue;
        }

        let results = esoteric_tests::run_all_tests(&data);
        let elapsed = t0.elapsed().as_secs_f64();
        let score = esoteric_tests::calculate_quality_score(&results);
        let passed = results.iter().filter(|r| r.passed).count();

        println!(
            " → {:.0}/100 ({}/{} passed) [{:.1}s]",
            score,
            passed,
            results.len(),
            elapsed
        );

        all_results.push((info.name.to_string(), data, results));
    }

    if all_results.is_empty() {
        eprintln!("No sources produced data.");
        std::process::exit(1);
    }

    // Generate report
    let report = generate_report(&all_results);

    if let Some(path) = output_path {
        std::fs::write(path, &report).unwrap();
        println!("\n📄 Report saved to: {path}");
    }

    // Summary table
    println!("\n{}", "=".repeat(60));
    println!(
        "{:<25} {:>6} {:>6} {:>8}",
        "Source", "Score", "Grade", "Pass"
    );
    println!("{}", "-".repeat(60));

    let mut sorted = all_results.clone();
    sorted.sort_by(|a, b| {
        let sa = esoteric_tests::calculate_quality_score(&a.2);
        let sb = esoteric_tests::calculate_quality_score(&b.2);
        sb.partial_cmp(&sa).unwrap()
    });

    for (name, _, results) in &sorted {
        let score = esoteric_tests::calculate_quality_score(results);
        let grade = if score >= 80.0 {
            'A'
        } else if score >= 60.0 {
            'B'
        } else if score >= 40.0 {
            'C'
        } else if score >= 20.0 {
            'D'
        } else {
            'F'
        };
        let passed = results.iter().filter(|r| r.passed).count();
        println!(
            "  {:<23} {:>5.1} {:>6} {:>4}/{}",
            name,
            score,
            grade,
            passed,
            results.len()
        );
    }
}

fn generate_report(results: &[(String, Vec<u8>, Vec<esoteric_tests::TestResult>)]) -> String {
    let mut report = String::new();
    report.push_str("# Esoteric Entropy — NIST Randomness Test Report\n\n");
    report.push_str(&format!("Generated: {}\n\n", chrono_now()));

    for (name, data, tests) in results {
        let score = esoteric_tests::calculate_quality_score(tests);
        let passed = tests.iter().filter(|r| r.passed).count();
        report.push_str(&format!("## {name}\n\n"));
        report.push_str(&format!(
            "- Samples: {} bytes\n- Score: {:.1}/100\n- Passed: {}/{}\n\n",
            data.len(),
            score,
            passed,
            tests.len()
        ));

        report.push_str("| Test | P | Grade | p-value | Statistic | Details |\n");
        report.push_str("|------|---|-------|---------|-----------|--------|\n");
        for t in tests {
            let ok = if t.passed { "✓" } else { "✗" };
            let pval = t
                .p_value
                .map(|p| format!("{p:.6}"))
                .unwrap_or_else(|| "—".to_string());
            report.push_str(&format!(
                "| {} | {} | {} | {} | {:.4} | {} |\n",
                t.name, ok, t.grade, pval, t.statistic, t.details
            ));
        }
        report.push_str("\n---\n\n");
    }

    report
}

fn chrono_now() -> String {
    // Simple timestamp without chrono dependency
    let dur = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default();
    format!("Unix timestamp: {}", dur.as_secs())
}
